#Importing all the libraries

In [1]:
import tensorflow as tf
import matplotlib.pyplot as plt
import numpy as np
from tensorflow.examples.tutorials.mnist import input_data

#Importing data
Mnist contains 28*28 pixels image
with output of 10 classes(0-9)
one_hot converts the output to a single value hot key
eg:- for 4 output will be [0 0 0 0 1 0 0 0 0 0]

In [2]:
mnist=input_data.read_data_sets('/temp/data',one_hot=True)

Instructions for updating:
Please use alternatives such as official/mnist/dataset.py from tensorflow/models.
Instructions for updating:
Please write your own downloading logic.
Instructions for updating:
Please use tf.data to implement this functionality.
Extracting /temp/data\train-images-idx3-ubyte.gz
Instructions for updating:
Please use tf.data to implement this functionality.
Extracting /temp/data\train-labels-idx1-ubyte.gz
Instructions for updating:
Please use tf.one_hot on tensors.
Extracting /temp/data\t10k-images-idx3-ubyte.gz
Extracting /temp/data\t10k-labels-idx1-ubyte.gz
Instructions for updating:
Please use alternatives such as official/mnist/dataset.py from tensorflow/models.


#Creating a function to create layers in neural network
Each layer takes input from previous layer and produces output to next layer
activation function is used to learn and make sense of the non-linear input data

In [3]:
def add_layer(inputs,in_size,out_size,activation=None):
    weight=tf.Variable(tf.random_normal([in_size,out_size]))
    bias=tf.Variable(tf.zeros([1,out_size])+0.1)
    lm=tf.matmul(inputs,weight)+bias
    if activation is None:
        output=lm
    else:
        output=activation(lm)
    return output

#Optimizer is used

In [4]:
xs=tf.placeholder(tf.float32,[None,784])
ys=tf.placeholder(tf.float32,[None,10])

output=add_layer(xs,784,10,tf.nn.softmax)
loss=tf.reduce_mean(-tf.reduce_sum(ys*tf.log(output),axis=1))
train=tf.train.GradientDescentOptimizer(0.5).minimize(loss)

#Session is initialised

In [5]:
sess=tf.InteractiveSession()
init=tf.global_variables_initializer()
sess.run(init)

#Finally model is trained

In [6]:
def accuracy_find(xs1,ys1):
    #outputs=sess.run(output,feed_dict={xs:xs1,ys:ys1})
    correct_prediction=tf.equal(tf.argmax(output,1),tf.argmax(ys1,1))
    accuracy=tf.reduce_mean(tf.cast(correct_prediction,tf.float32))
    print(sess.run(accuracy,feed_dict={xs:xs1,ys:ys1}))
    
for i in range(1000):
    batchx,batchy=mnist.train.next_batch(100)
    sess.run(train,feed_dict={xs:batchx,ys:batchy})
    if(i%50==0):
       accuracy_find(mnist.test.images,mnist.test.labels)
#accuracy_find(mnist.train.images,mnist.train.labels)
# sess.run(train,feed_dict={xs:mnist.train.images,ys:mnist.train.labels})
# accuracy_find(mnist.test.images,mnist.test.labels)


0.1072
0.6317
0.74
0.7809
0.7982
0.8207
0.8221
0.8391
0.8467
0.8501
0.8505
0.8569
0.8613
0.8618
0.8666
0.8626
0.8704
0.8679
0.8745
0.8708


0.1072
0.6317
0.74
0.7809
0.7982
0.8207
0.8221
0.8391
0.8467
0.8501
0.8505
0.8569
0.8613
0.8618
0.8666
0.8626
0.8704
0.8679
0.8745
0.8708
#So we can see the accuracy on test set gradually increases with trainning of next batches of train datas